src/evo/creditpool.h

```diff
@@ -0,0 +1,175 @@
+    {
+        READWRITE(
+            obj.locked,
+            obj.currentLimit,
+            obj.latelyUnlocked,
+            obj.indexes
+        );
+    }
+};
+
+/**
+ * The class CCreditPoolDiff has 2 purposes:
+ *  - it helps to determine which transaction can be included in new mined block
+ *  within current limits for Asset Unlock transactions and filter duplicated indexes
+ *  - to validate Asset Unlock transaction in mined block. The standalone checks of tx
+ *  such as CheckSpecialTx is not able to do so because at that moment there is no full
+ *  information about Credit Pool limits.
+ *
+ * CCreditPoolDiff temporary stores new values `lockedAmount` and `indexes` while
+ * limits should stay same and depends only on the previous block.
+ */
+class CCreditPoolDiff {
+private:
+    const CCreditPool pool;
+    std::unordered_set<uint64_t> newIndexes;
+
+    CAmount sessionLocked{0};
+    CAmount sessionUnlocked{0};
+
+    // target value is used to validate CbTx. If values mismatched, block is invalid
+    std::optional<CAmount> targetLocked;
+
+    const CBlockIndex *pindex{nullptr};
+public:
+    explicit CCreditPoolDiff(CCreditPool starter, const CBlockIndex *pindex, const Consensus::Params& consensusParams);
+
+    /**
+     * This function should be called for each Asset Lock/Unlock tx
+     * to change amount of credit pool
+     * @return true if transaction can be included in this block
+     */
+    bool processTransaction(const CTransaction& tx, TxValidationState& state);
+
+    CAmount getTotalLocked() const {
+        return pool.locked + sessionLocked - sessionUnlocked;
+    }
+
+    const std::optional<CAmount>& getTargetLocked() const {
+        return targetLocked;
+    }
+
+    std::string ToString() const {
+        return strprintf("CCreditPoolDiff(target=%lld,sessionLocked=%lld,sessionUnlocked=%lld,newIndexes=%lld,pool=%s", getTargetLocked() ? *getTargetLocked() : -1, sessionLocked, sessionUnlocked, newIndexes.size(), pool.ToString());
+    }
+
+private:
+    bool setTarget(const CTransaction& tx, TxValidationState& state);
+    bool lock(const CTransaction& tx, TxValidationState& state);
+    bool unlock(const CTransaction& tx, TxValidationState& state);
+};
+
+class CCreditPoolManager
+{
+private:
+    static constexpr size_t CreditPoolCacheSize = 1000;
+    RecursiveMutex cache_mutex;
+    unordered_lru_cache<uint256, CCreditPool, StaticSaltedHasher> creditPoolCache GUARDED_BY(cache_mutex) {CreditPoolCacheSize};
+
+    CEvoDB& evoDb;
+
+    static constexpr int DISK_SNAPSHOT_PERIOD = 576; // once per day
+
+public:
+    static constexpr int LimitBlocksToTrace = 576;
+    static constexpr CAmount LimitAmountLow = 100 * COIN;
+    static constexpr CAmount LimitAmountHigh = 1000 * COIN;
+
+    explicit CCreditPoolManager(CEvoDB& _evoDb);
+
+    ~CCreditPoolManager() = default;
+
+    /**
+      * @return CCreditPool with data or with empty depends on activation V19 at that block
```

Comment:
firstly, I just want to distinct for sure `not available` and `empty` values; because there's cache of credit pools in `creditPoolCache` and empty nullopt means that object is not found in cache. I feel like it is more error prune.

Secondly, even more important, since v20 activated - the value in CbTx should be always presented (even zero); before v20 - should not be even calculated.

So, empty and `nullopt` have different semantic meaning.